[ES|QL] Edits query in the dashboard

stratoula opened the pull request:
## Summary

Part of https://github.com/elastic/kibana/issues/165928
Closes https://github.com/elastic/kibana/issues/144498

Allows the user to edit the ES|QL query from the dashboard. Also allows the user to select one of the suggestions.

<img width="1886" alt="image" src="https://github.com/elastic/kibana/assets/17003240/9961c154-e414-4ce1-bff5-33ec5c30db69">
<img width="1883" alt="image" src="https://github.com/elastic/kibana/assets/17003240/6e8971d3-4a35-466f-804a-b8df58b09394">

### Testing
Navigate to Discover ES|QL mode and save a Lens chart to a dashboard. Click the edit Visualization.

### Important notes
- We can very easily enable suggestions for the dataview panels but I am going to do it on a follow up PR to keep this PR clean
- Creation is going to be added on a follow up PR
- Warnings are not rendered in the editor because I am using the limit 0 for performance reasons. We need to find another way to depict them via the embeddable or store. It will be on a follow up PR. 
- Errors are being displayed though. The user is not allowed to apply the changes when an error occurs.
- Creating ES|QL charts from dashboard will happen to a follow up PR

### Running queries which don't return numeric fields
In these cases (i.e. `from logstash-* | keep clientip` we are returning a table. I had to change the datatable logic for text based datasource to not depend to isBucketed flag. This is something we had foreseen from the [beginning of text based languages](https://github.com/elastic/kibana/issues/144498)  

<img width="1879" alt="image" src="https://github.com/elastic/kibana/assets/17003240/ca4b66fd-560d-4c1e-881d-b173458a06ae">

### Running queries which return a lot of fields
For queries with many fields Lens is going to suggest a huge table trying to add the fields to the different dimensions. This is not something we want:
- not performant
- user possibly will start removing fields from the dimensions
- this table is unreadable

For this reason we decided to select the first 5 fields and then the user can easily adjust the dimensions with the fields they want.

<img width="1215" alt="image" src="https://github.com/elastic/kibana/assets/17003240/07d7ee78-0085-41b1-98a0-a77eefbc0dcd">


### Checklist

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] Any UI touched in this PR is usable by keyboard only (learn more about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [ ] Any UI touched in this PR does not create any new axe failures (run axe in browser: [FF](https://addons.mozilla.org/en-US/firefox/addon/axe-devtools/), [Chrome](https://chrome.google.com/webstore/detail/axe-web-accessibility-tes/lhdoppojpmngadmnindnejefpokejbdd?hl=en-US))
- [ ] This renders correctly on smaller devices using a responsive layout. (You can test this [in your browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))
- [ ] This was checked for [cross-browser compatibility](https://www.elastic.co/support/matrix#matrix_browsers)

stratoula commented:
@elasticmachine merge upstream

stratoula commented on src/plugins/unified_histogram/public/layout/hooks/use_lens_suggestions.ts:
```diff
@@ -142,6 +144,7 @@ export const useLensSuggestions = ({
     currentSuggestion: histogramSuggestion ?? currentSuggestion,
     suggestionUnsupported: !currentSuggestion && !histogramSuggestion && isPlainRecord,
     isOnHistogramMode: Boolean(histogramSuggestion),
+    histogramQuery: histogramQuery.current ? { esql: histogramQuery.current } : undefined,
```
ℹ️  This is done to communicate correctly the histogram query (as it is not the same that Discover uses aka the unified search query)

davismcphee replied:
Thanks for the note 👍 Did it not work correctly before or do we just need to do this now because of the changes in this PR?

stratoula replied:
It did not work correctly before but it was not visible to the users as we didnt allow the editing of the query. So in 8.11 if the user saves the histogram to the dashboard they can edit it correctly but the query that is saved with the SO is not the correct one.

elasticmachine commented:
Pinging @elastic/kibana-visualizations (Team:Visualizations)

ninoslavmiskovic commented:
I experienced some different numbers when cloning a "Metric" visualization, but the query is the same. 

I will test more. 


https://github.com/elastic/kibana/assets/108192783/b409bc5e-20a0-4b3d-8dcc-d9508bc35c08

stratoula commented:
@ninoslavmiskovic this is happening due to the default 500 limit on the kibana side. As we explained on the offsite if you dont sort you might have different results. This is an ES behavior, it is not connected to my PR or any kibana related work :(

ninoslavmiskovic commented:
I understand. I thought it was only Discover but the challenge just got bigger here in my opinion, because duplicating charts is common and very visual. This just increases the sense of urgency on ES side to lift the limits.

mbondyra commented on packages/kbn-text-based-editor/src/text_based_languages_editor.tsx:
```diff
@@ -137,6 +166,7 @@ export const TextBasedLanguagesEditor = memo(function TextBasedLanguagesEditor({
   const [editorHeight, setEditorHeight] = useState(
     isCodeEditorExpanded ? EDITOR_INITIAL_HEIGHT_EXPANDED : EDITOR_INITIAL_HEIGHT
   );
+  const [hasSmallerSpace, setHasSmallerSpace] = useState(false);
```
nit: nonblocker, but how about changing the name to one of these: `isSpaceReduced` or `isLayoutCompact`, `isAreaReduced`

stratoula replied:
Good point dear, I will go with the isSpaceReduced I think

stratoula replied:
Done https://github.com/elastic/kibana/pull/169911/commits/25b28b6156f93162edd38741a426c030477001ac

mbondyra commented:
Such a great PR! 
I found one problem that I am not sure if related to this PR. The datatable config panel after choosing a field doesn't refresh unless something is changed or it's opened again.

https://github.com/elastic/kibana/assets/4283304/57ecb858-e648-481f-92c4-9507d4e5b32f

Also there's a problem with treemap:

https://github.com/elastic/kibana/assets/4283304/6c522ea4-254a-45c7-aecc-b5364c32878e

stratoula commented:
Thanx Marta! Treemap has a performance problem! We are tracking this here https://github.com/elastic/kibana/issues/164730 I will check the other and let you know!

stratoula commented:
@mbondyra the other one is an existing bug in main https://github.com/elastic/kibana/issues/171336. I created an issue to track it.

MichaelMarcialis reviewed (commented):
Thanks for putting this together, @stratoula. It's looking amazing, as usual from you! I've put together the following comments and questions after my initial review. Let me know if you have any questions.

## Flyout Header

- [ ] Can we update the flyout's header padding to be 16px on all sides to better match the designs? It looks like we're currently using 24px here instead.

- [ ] Let's change the "Edit visualization" text to "Edit ES|QL visualization". This better matches the designs and also makes it clearer to users that this visualization configuration differs from data view visualization configurations.

- [ ] It appears that the technical preview indicator is currently represented with an icon tooltip. Can we change this to use an icon-only beta badge with tooltip?

## Flyout Body

- [ ] I noticed an errant tick mark in the markup at the end of the flyout markup. I'm guessing this is just a typo somewhere in the code. Can we remove it?
  
  ![CleanShot 2023-11-15 at 15 33 35](https://github.com/elastic/kibana/assets/3884767/b920443a-92ae-46bb-be35-db660ac0082a)

### ES|QL Editor

- [ ] There appears to be a white gradient being rendered at the top of the ES|QL editor chrome. I'm guessing this is coming from the default flyout styles for when flyout body content is scrolling vertically (giving it that fading effect). Can we remove the gradient for this particular instance?
  
  ![CleanShot 2023-11-15 at 15 34 45](https://github.com/elastic/kibana/assets/3884767/727243e5-43c8-4188-af85-dbfa9b90ce14)

- [ ] The ES|QL editor header and footer were intended to be 48px tall. Currently they appear to be 34px and 42px respectively. Would it be possible to add the appropriate padding to each to appear at the intended height?

- [ ] The header and footer currently have left and right borders which cause them to look off or erroneous in some areas. Would it be possible to remove these borders to better match the designs?
  
  ![CleanShot 2023-11-15 at 15 35 55](https://github.com/elastic/kibana/assets/3884767/66deefa5-8548-418c-9055-5512d16248bd)

- [ ] The icon buttons in the header were intended to be 32x32px. Can we change them to be `size="s"`?

- [ ] It looks like the word wrapping option has been changed in the ES|QL editor from what was expected. The original design intent was that this would be a simple word wrap on/off setting. Instead, we're now offering a sort of automated formatting based on pipe usage (with word wrap being always on). Is this concept of a pipe wrapping a common one? I don't believe I've ever seen such a pattern in the wild, but I'll admit I'm probably not the typical audience for such a feature. If this is not a common pattern, I'd suggest we revert to the original design intent and change this option back to a pure word wrap on/off setting, with it turned on by default. Thoughts?

- [ ] The ES|QL editor body also appears to have borders on all sides. This causes the body to look odd on the left/right sides, and causes a doubling-of-border effect at the top, where it meets the header. Would it be possible to remove these borders?
  
  ![CleanShot 2023-11-15 at 15 36 49](https://github.com/elastic/kibana/assets/3884767/62311c0f-0c9c-4290-b0d4-3d97e74311f7)

- [ ] The original design intent for the "Run query" button was to keep it always enabled and let users rerun queries as often as they like. It looks like we're now disabling the run button here if no query changes are detected. Should we be disabling the "Run query" button when no query changes have been made? Is there a case for wanting to re-run a query when no changes have been made to the query? For example, what if the data set changes behind the scenes?

- [ ] Apologies in advance for my flip-flopping on this, but can we remove the prefixed `check` and `cross` icons from the timestamp messaging in the editor footer? I'm fine with this change affecting all ES|QL editors as well (i.e. unified search). I think the wording `@timestamp found` or `@timestamp not found` is sufficient, and the removal of these icons avoids us having yet another icon in the chrome, while preventing any potential confusion with error/warning messaging.
  
  ![CleanShot 2023-11-15 at 15 37 44](https://github.com/elastic/kibana/assets/3884767/39773af7-f194-4a30-a456-9167b30b3ce4)

- [ ] I've never really been a fan of the default EUI resizable button styles in these sorts of full-bleed panels, as it has a tendency to look odd and out-of-place. Can we override the styles or create our own version of the component that basically hides the initial default styles (the two stacked horizontal lines) altogether, and instead only shows the hover and active styles? I realize that will reduce discoverability of the resizing feature, but I'd prefer users to happily stumble across the feature as they move their mouse in the flyout, rather than show an ever-present eyesore. For what it's worth, I think the same direction can be taken for the resize button in the expanded unified search interface in ES|QL mode.
  
  ![CleanShot 2023-11-15 at 15 38 33](https://github.com/elastic/kibana/assets/3884767/b42f24fc-fee4-4690-95db-c7cdc7498ce7)

- [ ] After manually resizing the editor and then open/closing the accordions that follow, it appears that the editor bugs out and moves up a few pixels, cutting off half of the editor header. Would it be possible to fix this bug?
  
  ![CleanShot 2023-11-15 at 15 42 00](https://github.com/elastic/kibana/assets/3884767/44dbf994-9609-4d69-96d4-b45d716098b3)

### Accordions

- [ ] The original design intent for the accordions was that only one accordion could be open at a time, but both accordions can be closed if the user wished. Currently, it looks like we are correctly closing one accordion when the other is open. However, attempting to close both accordions does not appear to be possible. Instead, closing an accordion opens the opposing accordion. Can we fix this? The intended behavior can be viewed in the [Figma prototype](https://www.figma.com/proto/s62do8GDjERrRtp0sKKCJy/Mockups-%2F-Version-1-%5BIn-Progress%5D?page-id=&type=design&node-id=277-244134&viewport=-1307%2C-1232%2C0.5&t=fPnl24xnk6q2k6fn-1&scaling=min-zoom&starting-point-node-id=91%3A212667&mode=design).

- [ ] Opening and closing the accordions shows a brief moment where they jump up and over the editor space. Is there any way we can correct this visual bug?
  
  ![CleanShot 2023-11-15 at 15 44 08](https://github.com/elastic/kibana/assets/3884767/9edd93e5-c8e7-445b-bc80-c0b09d4c1d7e)

- [ ] Can we add a 1px border in between the "Layer configuration" and "Suggestions" accordions to better match the designs?
  
  ![CleanShot 2023-11-15 at 15 42 57](https://github.com/elastic/kibana/assets/3884767/8c87b68e-4cc0-4e1a-bfb0-5695181e7e21)

- [ ] Accordion buttons for "Layer configuration" and "Suggestions" should be 56px tall, with 16px padding on all sides to match designs. Can we add this in?

- [ ] The accordion contents for "Layer configuration" and "Suggestions" should also have 16px padding on the left, right, and bottom to match the designs.

- [ ] The background color for both "Layer configuration" and "Suggestions" is currently white. Can we change this to EUI's lightest shade to better match the designs?

#### Layer configuration

- [ ] Currently, the entire contents of the "Layer configuration" accordion is vertically scrollable. Can we change this to only allow the scrolling of layers panel(s), but not the top toolbar (so as to make the toolbar always visible and accessible)? Doing so would also better match the designs.
  
  ![CleanShot 2023-11-15 at 15 46 27](https://github.com/elastic/kibana/assets/3884767/955d297d-7a38-4196-8ce9-e08e5c719765)

- [ ] As ES|QL doesn't use data views and since we are currently only supporting a single layer for ES|QL-based visualizations for the time being, the original designs suggested removing the data view selector for layer panel altogether. Here I see we still show them, but in a disabled state. Can we remove them altogether until we determine how we're handling multiple layers in ES|QL-based visualizations?
  
  ![CleanShot 2023-11-15 at 15 47 24](https://github.com/elastic/kibana/assets/3884767/d09a07c9-98ef-482b-a2c7-bf00a3b86abd)

#### Suggestions

- [ ] The "Suggestions" accordion button should include a right-aligned counter badge indicating how many suggestions are available when the accordion is closed. Possible to add?
  
  ![CleanShot 2023-11-15 at 15 49 25](https://github.com/elastic/kibana/assets/3884767/e25bbd7b-37d1-4985-bc1a-eaaac5ac3456)

- [ ] Suggestion panels should have 16px gap between each other, both vertically and horizontally. Can we add this in to better match the designs?

## Flyout Footer

- [ ] Similar to the "Run query" button, the original designs suggested always keeping the "Apply and close" button enabled, even if no changes had been made. To my mind it doesn't hurt the experience to keep it enabled. Thoughts? Is there a compelling reason to disable?

davismcphee commented on src/plugins/unified_histogram/public/layout/hooks/use_lens_suggestions.ts:
```diff
@@ -61,7 +61,7 @@ export const useLensSuggestions = ({
   const currentSuggestion = originalSuggestion ?? suggestions.firstSuggestion;
   const suggestionDeps = useRef(getSuggestionDeps({ dataView, query, columns }));
-
+  const histogramQuery = useRef<string>('');
```
```suggestion
  const histogramQuery = useRef<AggregateQuery>();
```
Could we store the whole query object in the ref instead to avoid returning a new object reference on each render? Just thinking it could trigger some unwanted effects given the memoization we use elsewhere in Unified Histogram.

stratoula replied:
Thanx my dear Davis. Addressed here https://github.com/elastic/kibana/pull/169911/commits/d94f1a481670ea70f0ea6598bd38631c9481a5af

davismcphee reviewed (commented):
Quick initial code review of Data Discovery changes.

stratoula commented:
@MichaelMarcialis thanx a lot for your thorough review as always! It will take me some time to address everything but I will ping you to review again when I feel is ready. 🙇

mbondyra commented on src/plugins/unified_histogram/public/layout/hooks/use_lens_suggestions.ts:
```diff
@@ -107,15 +106,18 @@ export const useLensSuggestions = ({
+        histogramQuery.current = { esql: esqlQuery };
         return sug[0];
       }
+      histogramQuery.current = undefined;
       return undefined;
```
these two lines are not necessary, are they? it would just follow line 120 and 121

stratoula replied:
I will remove them

mbondyra commented on x-pack/plugins/lens/public/app_plugin/shared/edit_on_the_fly/flyout_wrapper.tsx:
```diff
@@ -0,0 +1,153 @@
+              <EuiFlexGroup alignItems="center" gutterSize="xs">
+                <EuiFlexItem grow={false}>
+                  <EuiTitle size="xs">
+                    <h2 id="Edit visualization">
```
why do we add id that is a white-spaced string? Any reason for it?

stratoula replied:
No reason I just took an example from EUI, I will remove it.

mbondyra reviewed (approved):
Tested on Chrome, left some small comments but code LGTM 👌🏼

MichaelMarcialis commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/suggestion_panel.tsx:
```diff
@@ -474,64 +514,92 @@ export function SuggestionPanel({
+      buttonContent={buttonContent}
+      forceState={hideSuggestions ? 'closed' : 'open'}
+      onToggle={toggleSuggestions}
+      extraAction={
+        existsStagedPreview &&
+        !hideSuggestions && (
+          <EuiToolTip
+            content={i18n.translate('xpack.lens.suggestion.refreshSuggestionTooltip', {
+              defaultMessage: 'Refresh the suggestions based on the selected visualization.',
+            })}
+          >
+            <EuiButtonEmpty
+              data-test-subj="lensSubmitSuggestion"
+              size="xs"
+              iconType="refresh"
+              onClick={() => {
+                dispatchLens(submitSuggestion());
+              }}
+            >
+              {i18n.translate('xpack.lens.sugegstion.refreshSuggestionLabel', {
+                defaultMessage: 'Refresh',
+              })}
+            </EuiButtonEmpty>
+          </EuiToolTip>
+        )
+      }
```
This may be partially tangential (as it also affects the Lens app), but I wanted to mention it here because it's something that has always bugged me. I've never been a fan of the "Refresh" button that appears when users select one of the available suggestions. I understand we probably included it because Lens have no formal "undo" function, and this "Refresh" button provides a means users to either confirm adoption of the newly selected suggestion or to go back to their previous chart configuration. 

What if instead of showing this confirming "Refresh" button, we assume any newly selected suggestion is not adopted until users either 1) make a configuration change to their visualization or 2) save their visualization in Lens (or apply their changes in the Dashboard inline visualization editor). If one of those occur, it is assumed the suggestion has been adopted by the user and the suggestions can be refreshed. If those don't occur, the user can still return to their previous state by clicking the original suggestion tile. We can also change the "Current visualization" text to the appropriate visualization type with this approach. 

Thoughts? CCing @timductive and @ninoslavmiskovic in case they have any concerns about such a change. That said, if we don't want such a change at this time, can we change the placement of this "Refresh" button so that it appears before the suggestion count, rather than after?

![CleanShot 2023-11-21 at 15 08 55](https://github.com/elastic/kibana/assets/3884767/381015db-2e35-4ab0-8760-b2d11e83dd88)

stratoula replied:
I fixed the order! Let's discuss the other suggestion on a separate topic as it is not related with the current change!

MichaelMarcialis reviewed (approved):
Thanks so much for those changes and comments, @stratoula! I've added my responses below, as well as a few new comments for your re-review. 

I'm going to be out on PTO this Thursday and Friday for Thanksgiving in the US and all of next week for the UX Summit. I know you're also out until this Wednesday for ElasticON. Given those circumstances, I'm going to go ahead and approve this now, under the assumption that you're able to address my comments (in the hopes that I don't hold you up for a week+). Thanks!

## Response to previous comments

> > It looks like the word wrapping option has been changed in the ES|QL editor from what was expected. The original design intent was that this would be a simple word wrap on/off setting. Instead, we're now offering a sort of automated formatting based on pipe usage (with word wrap being always on). Is this concept of a pipe wrapping a common one? I don't believe I've ever seen such a pattern in the wild, but I'll admit I'm probably not the typical audience for such a feature. If this is not a common pattern, I'd suggest we revert to the original design intent and change this option back to a pure word wrap on/off setting, with it turned on by default. Thoughts?
> 
> This was a feature request. Many customers asked to make it work like that which makes sense to me. It is very useful to wrap and unwrap by pipes so I would like to keep it as it is

Thanks for that context. Given that customers have specifically requested such functionality, I'm not opposed to its inclusion as long as we can make a few small tweaks to accommodate for the differences between it and the previous word-wrapping:

- Rather than reusing the word wrap icons here, perhaps we should create new icons that better represent the addition or removal of line breaks on pipes. If you agree, I'd be happy to make some.

- The boolean-type button (on/off) feels strange in practice. While it made sense for word-wrap to be on a single toggling button, it gets a little muddy with the concept of formatting the query to add or remove line breaks on pipes. It's not a true boolean setting. For example, even if "Wrap with pipes" is disabled, I can still manually format all or part of my query to have line breaks before my pipes. The verbiage may also be contributing to this strangeness, as the action is technically more akin to adding/removing line breaks than "wrapping". To remedy this, would you be open to replacing the single boolean button with two ever-present buttons that allow the user to "Add line breaks on pipes" and "Remove line breaks on pipes"?

Thoughts? If you agree, let me know and I can work up some new icons when I get a moment and we can split this off as a separate issue.

> > The original design intent for the "Run query" button was to keep it always enabled and let users rerun queries as often as they like. It looks like we're now disabling the run button here if no query changes are detected. Should we be disabling the "Run query" button when no query changes have been made? Is there a case for wanting to re-run a query when no changes have been made to the query? For example, what if the data set changes behind the scenes?
> 
> I think at this case they will run the Refresh button which is what they already do for the rest of the panels. It is more performant like that, the user wont click it without a reason.

Given the better performance of the unified search refresh button that you mentioned, would it make sense to change the "Run query" button in the edit visualization flyout to a "Refresh" button, similar to how we do in unified search? That way there is a refresh button local to the section the user is working in (rather than disabling the "Run query" button and asking them to go to a different section on the page to refresh)?

> > I've never really been a fan of the default EUI resizable button styles in these sorts of full-bleed panels, as it has a tendency to look odd and out-of-place. Can we override the styles or create our own version of the component that basically hides the initial default styles (the two stacked horizontal lines) altogether, and instead only shows the hover and active styles? I realize that will reduce discoverability of the resizing feature, but I'd prefer users to happily stumble across the feature as they move their mouse in the flyout, rather than show an ever-present eyesore. For what it's worth, I think the same direction can be taken for the resize button in the expanded unified search interface in ES|QL mode.
> 
> I am fine with that but I prefer to address it in a separate PR. I created an ER here #171379.

Sounds good. Thanks for opening the issue!

> > Currently, the entire contents of the "Layer configuration" accordion is vertically scrollable. Can we change this to only allow the scrolling of layers panel(s), but not the top toolbar (so as to make the toolbar always visible and accessible)? Doing so would also better match the designs.
> 
> The best I can do is to change the toolbar to fixed position.
> 
> If you are ok with it I can push the code. Otherwise I will need some help.

Based on this GIF, I think that works. The only things to note are:

- It looks like the left edge shadow of the layer panel is overflowing its scrolling container when the layer panel moves up and out of sight, causing the shadow to then appear behind the toolbar buttons. Would this be possibly to remedy?

- It would be great if we could apply the CSS gradient fade effect (that EUI uses on some of its borderless scrolling containers) at the top of the scroll container.

> > After manually resizing the editor and then open/closing the accordions that follow, it appears that the editor bugs out and moves up a few pixels, cutting off half of the editor header. Would it be possible to fix this bug?
> 
> I can't reproduce it. Can you give me more detailed steps?

It looks like this has been resolved, as I can no longer reproduce it either. Thanks for looking into it!

> > Opening and closing the accordions shows a brief moment where they jump up and over the editor space. Is there any way we can correct this visual bug?
> 
> I can reproduce it but I don't know how to fix it. I will keep looking into it but maybe you have an idea? It is due to the CSS we added for the accordions scroll but I am not sure which specific property creates it.

Let's hold off on this one until some of the markup and styles refactoring that I mention below are implemented, as those changes will likely end up impacting how we decide to tackle this visual bug. We can probably split it off as a separate issue for the time being.

## New comments

- [ ] If I edit my ES|QL query in the edit visualization flyout and click "Apply and close" before clicking "Run query", none of my query edits are run or saved. Instead, if it's detected that a query change has been made but not yet run, can we have it so that "Apply and close" also run those query changes?

- [ ] I noticed that there was a little gray dot appearing in the top-left of the editor chrome. It appears to be caused by one of the tooltip elements for the Monaco editor. Is it possible for us to hide or remove this?
  
  ![CleanShot 2023-11-21 at 13 39 16](https://github.com/elastic/kibana/assets/3884767/5c80b4e7-96aa-4ace-87ad-0e32d1414139)

- [ ] It appears that the height of the "Suggestions" accordion header is correct at 56px, but the height of the "Layer configuration" accordion header is 52px. Can we make both 56px for consistency?

- [ ] The padding to the left and right of the accordion header buttons doesn't trigger the accordion. Would it be possible to restructure the markup and styles so that the horizontal padding is part of the button that triggers the accordion (rather than on the container that houses that button)? That way users can click anywhere in those accordion header boxes to expand/contract the accordion contents. Happy to help on the HTML/CSS side if needed.
  
  ![CleanShot 2023-11-21 at 16 19 50](https://github.com/elastic/kibana/assets/3884767/44c71dfd-2541-4bc6-abdd-76813c3eee0f)

- [ ] Similar to the above, the way the markup and styles are currently configured, the "Layer configuration" accordion contents have the scrollbar being kicked in by the right padding. Can we change it so that the padding is applied directly to the element that renders the scrollbar, so as to make the padding appear between the content and scrollbar instead? I imagine the same would need to be done to the "Suggestion" accordion contents as well.
  
  ![CleanShot 2023-11-21 at 16 21 30](https://github.com/elastic/kibana/assets/3884767/bfc2b14d-55b9-4231-95ee-5e4cf3ac29f2)

- [ ] Can we remove the extraneous `EuiSpacer` under the layer panel visualization type selector?
  
  ![CleanShot 2023-11-21 at 16 22 07](https://github.com/elastic/kibana/assets/3884767/56635a05-6284-47e8-a1ff-ab022e968251)

- [ ] The suggestion panel buttons are currently using a right margin of 16px to create the horizontal space between each suggestion. Unfortunately, this is also causing a space of 32px between the suggestions and the rightmost edge of their container. Can we remedy this? I imagine this can be achieved by either placing a -16px right margin on their container or using the `gap` property for flex spacing, instead of margins.
  
  ![CleanShot 2023-11-21 at 16 07 30](https://github.com/elastic/kibana/assets/3884767/80abae2b-5b2e-485b-97c7-aae444d72ca5)

davismcphee reviewed (approved):
Awesome work and such a cool feature! 🙌 Approving now because the Data Discovery/Unified Histogram changes LGTM, but I noticed a few things while testing that might be worth looking into or following up on later.

It looks like we add a visualization title by default in Discover when saving an ES|QL viz, but it disappears if I dismiss the modal and reopen it (this may have already existed before this PR, but figured it was worth noting):

https://github.com/elastic/kibana/assets/25592674/24909242-411d-48c6-9f4e-151adb9ceea4

When I run the ES|QL query in the Dashboard flyout, it changes my selected viz type. Maybe this is expected, but it surprised me so I figured I'd mention it. In this case I modified the query, but it happens even if I don't change it and just run the query with `⌘⏎`:

https://github.com/elastic/kibana/assets/25592674/4ab98e35-ae62-48c4-8a20-69c481aac84b

Running the query resets the ES|QL panel height after resizing it:

https://github.com/elastic/kibana/assets/25592674/b75884da-a681-42d5-b923-27146c85e30f

Clicking the refresh button in the suggestions panel after selecting a suggestion causes the section to disappear. Maybe this is just because of the rotating number example viz and not a real scenario in prod?

https://github.com/elastic/kibana/assets/25592674/c075a8e3-c921-4014-85c0-ccc2fc544019

stratoula commented on x-pack/plugins/lens/public/editor_frame_service/editor_frame/suggestion_panel.tsx:
```diff
@@ -474,64 +514,92 @@ export function SuggestionPanel({
-          data-test-subj="lnsSuggestionsPanel"
-          role="list"
-          tabIndex={0}
+          css={css`
+            background-color: ${euiThemeVars.euiColorLightShade};
+            height: ${euiThemeVars.euiSizeL};
+            width: ${euiThemeVars.euiSizeL};
+            text-align: center;
+            border-radius: ${euiThemeVars.euiSizeXS};
+          `}
         >
-          {changesApplied ? renderSuggestionsUI() : renderApplyChangesPrompt()}
+          {suggestions.length + 1}
         </div>
```
Done!

stratoula commented:
@MichaelMarcialis 

> Thoughts? If you agree, let me know and I can work up some new icons when I get a moment and we can split this off as a separate issue.

Absolutely! If you provide me the icons I will address this to a separate PR. Also maybe we can improve the wording here (?)

> Based on this GIF, I think that works. The only things to note are ...

I will create a new issue for this and try to improve this on a separate PR. Not sure if addressing your comments will be trivial and I want to merge this PR

> would it make sense to change the "Run query" button in the edit visualization flyout to a "Refresh" button, similar to how we do in unified search? 

Not sure about it. 🤔 Sometimes it can refresh and sometimes it can run a complete new query. Maybe we can detect if it refreshes or not and change the button text? Let's discuss when you are back and address this to a separate PR.

> If I edit my ES|QL query in the edit visualization flyout and click "Apply and close" before clicking "Run query", none of my query edits are run or saved. Instead, if it's detected that a query change has been made but not yet run, can we have it so that "Apply and close" also run those query changes?

I see what you say but this is not as trivial as it seems. I want to think about it. For now I will only allow the Apply to run if changes have been made

> The padding to the left and right of the accordion header buttons doesn't trigger the accordion. Would it be possible to restructure the markup and styles so that the horizontal padding is part of the button that triggers the accordion (rather than on the container that houses that button)? That way users can click anywhere in those accordion header boxes to expand/contract the accordion contents. Happy to help on the HTML/CSS side if needed.

I could not make it work. Moving the padding to the button css doesnt respect the arrow so yes I would like some help here!

<img width="478" alt="image" src="https://github.com/elastic/kibana/assets/17003240/ed191c13-df17-44b1-a108-ab277eb39d9b">

> Similar to the above, the way the markup and styles are currently configured, the "Layer configuration" accordion contents have the scrollbar being kicked in by the right padding. Can we change it so that the padding is applied directly to the element that renders the scrollbar, so as to make the padding appear between the content and scrollbar instead? I imagine the same would need to be done to the "Suggestion" accordion contents as well.

Happy to do that but I am not sure how to do that tbh. I will create an issue and we can look into it together when you are back.

stratoula commented:
@davismcphee 

> It looks like we add a visualization title by default in Discover when saving an ES|QL viz, but it disappears if I dismiss the modal and reopen it (this may have already existed before this PR, but figured it was worth noting):

Correct, this is not related with the changes addressed in this PR. Please create an ER if you want this for Discover. I dont have a strong opinion


> When I run the ES|QL query in the Dashboard flyout, it changes my selected viz type. Maybe this is expected, but it surprised me so I figured I'd mention it. In this case I modified the query, but it happens even if I don't change it and just run the query with ⌘⏎:

Expected for now, I guess we could store the selection on the url state but is not planned for now and is unerelated with this PR


> Running the query resets the ES|QL panel height after resizing it:

Expected for now but we can def improve. I will create an ER

> Clicking the refresh button in the suggestions panel after selecting a suggestion causes the section to disappear. Maybe this is just because of the rotating number example viz and not a real scenario in prod?

Correct, you wont see this in production!

ninoslavmiskovic commented:
> When I try to save the default histogram visualization and edit the query on a dashboard, I see the underlining ES|QL query that I guess is used to create the histogram in Discover. 
> 
> 
> 
> This could confuse the user since the queries are not the same. - The query on the dashboard make sense, just might cause some confusion that the user does not see the same query. 
> 
> 
> 
> https://github.com/elastic/kibana/assets/108192783/23f64765-b619-4794-94cd-e5cdced5181e
> 
> 
> 
>  

@stratoula did you have time to look at above 👆

stratoula commented:
@ninoslavmiskovic the correct query for the histogram is the one you see in the dashboard. I think that an advanced user understands that the `from index` doesnt create time buckets and that we are using a different query :D They can also see it on the inspector :D.

I am not sure what we can do here. If the users find it confusing maybe we can discuss with the UX team but it is wrong to use the Discover unified search query for this chart if this is what you imply.

ninoslavmiskovic commented:
@stratoula . We have had some users who do not quite understand the difference between the two table modes we have in Discover, and this challenge is kind of connected because it is only present if the user saves the histogram without doing STATS in Discover.

I am not applying to use the Discover query, because as you are stating it is wrong. I am just raising a potential UX challenge. 

The challenge gets more visible if the user has evolved the query with e.g. a couple of "WHERE" and "EVAL"s 

**See example:**  

https://github.com/elastic/kibana/assets/108192783/64e15048-0106-4131-a1de-3bb5e92e9f0f

The query is exactly the same if the user has aggregations and then saves the visualization, but not when they are not, then we are actually also adding a "STATS" the the query.  

There is just some inconsistency in the flow. E.g. If you saved the query as a saved search, and you go back and forward you see the same query, but if you save the histogram, then you don't. 

https://github.com/elastic/kibana/assets/108192783/8c4908b2-cfd8-49c7-8ac3-edf1cc6f641e

Showing the full query in Discover that is powering the visualization would not work. 

I agree let's hear some UX folks on some input  @MichaelMarcialis

stratoula commented:
> There is just some inconsistency in the flow. E.g. If you saved the query as a saved search, and you go back and forward you see the same query, but if you save the histogram, then you don't.

This is why the saved searches should be renamed differently because as I have said in the past you save the Discover view not only the query

> We have had some users who do not quite understand the difference between the two table modes we have in Discover,

You mean on the dataview mode? Because both in dataview mode and ES|QL we have the same concept, different queries create the histogram.

Anyway I feel this discussion is irrelevant with this PR :D

ninoslavmiskovic commented:
@stratoula  Sure, we can take the discussion in a meeting or so. 👌🏻  For me, it was just important to note that users who go through this flow will not get 1:1 queries as they will when using aggregations starting in Discover.

MichaelMarcialis commented:
Thanks for the responses, @stratoula. Please forgive my delayed reply, as I was traveling last week. 

> > Thoughts? If you agree, let me know and I can work up some new icons when I get a moment and we can split this off as a separate issue.
> 
> Absolutely! If you provide me the icons I will address this to a separate PR. Also maybe we can improve the wording here (?)

Sounds good. I'll try to work up some icon and wording suggestions for you this week.

> > Based on this GIF, I think that works. The only things to note are ...
> 
> I will create a new issue for this and try to improve this on a separate PR. Not sure if addressing your comments will be trivial and I want to merge this PR

Thank you!

> > would it make sense to change the "Run query" button in the edit visualization flyout to a "Refresh" button, similar to how we do in unified search?
> 
> Not sure about it. 🤔 Sometimes it can refresh and sometimes it can run a complete new query. Maybe we can detect if it refreshes or not and change the button text? Let's discuss when you are back and address this to a separate PR.

Exactly. If we'd be able to enhance this button to detect and conditionally show as "Run query" or "Refresh", that would be great. Let me know if you want to discuss further or want me to write up an issue.

> > If I edit my ES|QL query in the edit visualization flyout and click "Apply and close" before clicking "Run query", none of my query edits are run or saved. Instead, if it's detected that a query change has been made but not yet run, can we have it so that "Apply and close" also run those query changes?
> 
> I see what you say but this is not as trivial as it seems. I want to think about it. For now I will only allow the Apply to run if changes have been made

When you say "I will only allow the Apply to run if changes have been made", I think that's what I'm asking for, as it didn't happen that way in my testing. When "Apply and close" is clicked and the current query has been changed but not run, then "Apply and close" will also save the query changes and run them. Or are you speaking about something different?

> > The padding to the left and right of the accordion header buttons doesn't trigger the accordion. Would it be possible to restructure the markup and styles so that the horizontal padding is part of the button that triggers the accordion (rather than on the container that houses that button)? That way users can click anywhere in those accordion header boxes to expand/contract the accordion contents. Happy to help on the HTML/CSS side if needed.
> 
> I could not make it work. Moving the padding to the button css doesnt respect the arrow so yes I would like some help here!
> 
> <img alt="image" width="478" src="https://private-user-images.githubusercontent.com/17003240/285121992-ed191c13-df17-44b1-a108-ab277eb39d9b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE3MjEyMDcsIm5iZiI6MTcwMTcyMDkwNywicGF0aCI6Ii8xNzAwMzI0MC8yODUxMjE5OTItZWQxOTFjMTMtZGYxNy00NGIxLWExMDgtYWIyNzdlYjM5ZDliLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA0VDIwMTUwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVhN2EwZjVlNmUyOGE2OGZlNTI4YWZmMjcxZjU3M2ZhMGM5OGYyODhkOTA3NTJjZjIxNGVkYTM0ZTM4MjgxMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.nh_XEMjIySHGBXvj5Vhy6DCSSbTSQEJshsSYNLUVBgI">
> > Similar to the above, the way the markup and styles are currently configured, the "Layer configuration" accordion contents have the scrollbar being kicked in by the right padding. Can we change it so that the padding is applied directly to the element that renders the scrollbar, so as to make the padding appear between the content and scrollbar instead? I imagine the same would need to be done to the "Suggestion" accordion contents as well.
> 
> Happy to do that but I am not sure how to do that tbh. I will create an issue and we can look into it together when you are back.

Sounds good. I'll throw some time on the calendar for us to do some CSS pairing.

stratoula commented:
@MichaelMarcialis about this:

> When you say "I will only allow the Apply to run if changes have been made", I think that's what I'm asking for, as it didn't happen that way in my testing. When "Apply and close" is clicked and the current query has been changed but not run, then "Apply and close" will also save the query changes and run them. Or are you speaking about something different?

I don't think it is a good idea to run the query when the apply button is clicked. The query might be problematic or not produce a valid suggestion. The current experience makes more sense to me as a user. It tries to not produce problematic panels which I think is the correct experience here. But maybe you are thinking it differently. Let's discuss it synchronously. I have my reservations but happy to hear your thoughts.

MichaelMarcialis commented:
> I don't think it is a good idea to run the query when the apply button is clicked. The query might be problematic or not produce a valid suggestion. The current experience makes more sense to me as a user. It tries to not produce problematic panels which I think is the correct experience here. But maybe you are thinking it differently. Let's discuss it synchronously. I have my reservations but happy to hear your thoughts.

Yeah, let's plan to talk about it in our Friday meeting. I can appreciate your perspective that you feel running a changed query might be too heavy handed "Apply and close" is clicked. My concern is less about the query running and more of the query changes being saved/stored when "Apply and close" is clicked. To me, it feels strange to not have my query changes saved/stored when I use the apply action. Unfortunately in this case, saving/storing those changes also necessitates running the query, given that the inline editor is being dismissed in the process.
